File: examples/python/product/product_logic.py
"""Product command handlers and business logic."""

from datetime import datetime, timezone

import structlog
from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains


class CommandRejectedError(Exception):
    """Command was rejected due to business rule violation."""


def handle_create_product(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.ProductState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle CreateProduct command."""
    if state.sku:
        raise CommandRejectedError("Product already exists")

    cmd = domains.CreateProduct()
    command_any.Unpack(cmd)

    if not cmd.sku:
        raise CommandRejectedError("SKU is required")
    if not cmd.name:
        raise CommandRejectedError("Name is required")
    if cmd.price_cents < 0:
        raise CommandRejectedError("Price cannot be negative")

    log.info("creating_product", sku=cmd.sku, name=cmd.name, price_cents=cmd.price_cents)

    event = domains.ProductCreated(
        sku=cmd.sku,
        name=cmd.name,
        description=cmd.description,
        price_cents=cmd.price_cents,
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )


def handle_update_product(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.ProductState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle UpdateProduct command."""
    if not state.sku:
        raise CommandRejectedError("Product does not exist")
    if state.status == "discontinued":
        raise CommandRejectedError("Cannot update discontinued product")

    cmd = domains.UpdateProduct()
    command_any.Unpack(cmd)

    if not cmd.name:
        raise CommandRejectedError("Name is required")

    log.info("updating_product", name=cmd.name)

    event = domains.ProductUpdated(
        name=cmd.name,
        description=cmd.description,
        updated_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )


def handle_set_price(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.ProductState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle SetPrice command."""
    if not state.sku:
        raise CommandRejectedError("Product does not exist")
    if state.status == "discontinued":
        raise CommandRejectedError("Cannot change price of discontinued product")

    cmd = domains.SetPrice()
    command_any.Unpack(cmd)

    if cmd.new_price_cents < 0:
        raise CommandRejectedError("Price cannot be negative")

    log.info("setting_price", old_price=state.price_cents, new_price=cmd.new_price_cents)

    event = domains.PriceSet(
        new_price_cents=cmd.new_price_cents,
        previous_price_cents=state.price_cents,
        set_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )


def handle_discontinue(
    command_book: angzarr.CommandBook,
    command_any: Any,
    state: domains.ProductState,
    seq: int,
    log: structlog.BoundLogger,
) -> angzarr.EventBook:
    """Handle Discontinue command."""
    if not state.sku:
        raise CommandRejectedError("Product does not exist")
    if state.status == "discontinued":
        raise CommandRejectedError("Product already discontinued")

    cmd = domains.Discontinue()
    command_any.Unpack(cmd)

    log.info("discontinuing_product", reason=cmd.reason)

    event = domains.ProductDiscontinued(
        reason=cmd.reason,
        discontinued_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[
            angzarr.EventPage(
                num=seq,
                event=event_any,
                created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
            )
        ],
    )

File: examples/csharp/Product/src/ProductLogic.cs
using Angzarr;
using Examples;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Serilog;

namespace Angzarr.Examples.Product;

public class ProductLogic : IProductLogic
{
    private readonly Serilog.ILogger _logger;

    public ProductLogic(Serilog.ILogger logger)
    {
        _logger = logger.ForContext<ProductLogic>();
    }

    public ProductState RebuildState(EventBook? eventBook)
    {
        var state = ProductState.Empty;

        if (eventBook == null || eventBook.Pages.Count == 0)
            return state;

        foreach (var page in eventBook.Pages)
        {
            if (page.Event == null) continue;
            state = ApplyEvent(state, page.Event);
        }

        return state;
    }

    private ProductState ApplyEvent(ProductState state, Any eventAny)
    {
        var typeUrl = eventAny.TypeUrl;

        if (typeUrl.EndsWith("ProductCreated"))
        {
            var evt = eventAny.Unpack<ProductCreated>();
            return new ProductState(evt.Sku, evt.Name, evt.Description, evt.PriceCents, "active");
        }

        if (typeUrl.EndsWith("ProductUpdated"))
        {
            var evt = eventAny.Unpack<ProductUpdated>();
            return state with { Name = evt.Name, Description = evt.Description };
        }

        if (typeUrl.EndsWith("PriceSet"))
        {
            var evt = eventAny.Unpack<PriceSet>();
            return state with { PriceCents = evt.NewPriceCents };
        }

        if (typeUrl.EndsWith("ProductDiscontinued"))
        {
            return state with { Status = "discontinued" };
        }

        return state;
    }

    public EventBook HandleCreateProduct(ProductState state, string sku, string name, string description, int priceCents)
    {
        if (state.Exists)
            throw CommandValidationException.FailedPrecondition("Product already exists");

        if (string.IsNullOrWhiteSpace(sku))
            throw CommandValidationException.InvalidArgument("SKU is required");

        if (string.IsNullOrWhiteSpace(name))
            throw CommandValidationException.InvalidArgument("Product name is required");

        if (priceCents <= 0)
            throw CommandValidationException.InvalidArgument("Price must be positive");

        _logger.Information("creating_product {@Data}", new { sku, name, price_cents = priceCents });

        var evt = new ProductCreated
        {
            Sku = sku,
            Name = name,
            Description = description ?? "",
            PriceCents = priceCents,
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleUpdateProduct(ProductState state, string name, string description)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Product does not exist");

        if (state.IsDiscontinued)
            throw CommandValidationException.FailedPrecondition("Cannot update discontinued product");

        if (string.IsNullOrWhiteSpace(name))
            throw CommandValidationException.InvalidArgument("Product name is required");

        _logger.Information("updating_product {@Data}", new { sku = state.Sku, name });

        var evt = new ProductUpdated
        {
            Name = name,
            Description = description ?? ""
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleSetPrice(ProductState state, int priceCents)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Product does not exist");

        if (state.IsDiscontinued)
            throw CommandValidationException.FailedPrecondition("Cannot set price on discontinued product");

        if (priceCents <= 0)
            throw CommandValidationException.InvalidArgument("Price must be positive");

        _logger.Information("setting_price {@Data}", new { sku = state.Sku, old_price = state.PriceCents, new_price = priceCents });

        var evt = new PriceSet
        {
            OldPriceCents = state.PriceCents,
            NewPriceCents = priceCents
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleDiscontinue(ProductState state)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Product does not exist");

        if (state.IsDiscontinued)
            throw CommandValidationException.FailedPrecondition("Product already discontinued");

        _logger.Information("discontinuing_product {@Data}", new { sku = state.Sku });

        var evt = new ProductDiscontinued
        {
            DiscontinuedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    private static EventBook CreateEventBook(IMessage evt)
    {
        var page = new EventPage
        {
            Num = 0,
            Event = Any.Pack(evt),
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        var book = new EventBook();
        book.Pages.Add(page);
        return book;
    }
}
